agent/system_metrics.go

```diff
@@ -0,0 +1,256 @@
+	s, err := systemMetricsCollectorSetup(ctx, interval, t, collectors)
+	if err != nil {
+		return nil, err
+	}
+
+	mc, err := metrics.NewSystemMetricsClientWithExistingConnection(ctx, conn)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem creating new system metrics client")
+	}
+	s.client = mc
+	return s, nil
+}
+
+func systemMetricsCollectorSetup(ctx context.Context, interval time.Duration, t *task.Task,
+	collectors []MetricCollector) (*SystemMetricsCollector, error) {
+	if interval < 0 {
+		return nil, errors.New("interval cannot be negative")
+	}
+	if interval == 0 {
+		interval = time.Minute
+	}
+
+	taskOpts, err := getTaskInfo(t)
+	if err != nil {
+		return nil, errors.Wrap(err, "incomplete task metadata")
+	}
+
+	if len(collectors) == 0 {
+		return nil, errors.New("must provide at least one MetricCollector")
+	}
+
+	return &SystemMetricsCollector{
+		interval:   interval,
+		collectors: collectors,
+		taskOpts:   taskOpts,
+		catcher:    grip.NewBasicCatcher(),
+	}, nil
+}
+
+// Start commences collecting metrics using each of the provided MetricCollectors.
+// Regardless of if Start returns an error, Close should still be called
+// to close any opened connections, set the exit code, and handle any returned
+// errors. This can also be handled by cancelling the provided context, but
+// any errors will only be logged to the global error logs in this case.
+func (s *SystemMetricsCollector) Start(ctx context.Context) error {
+	var err error
+	s.id, err = s.client.CreateSystemMetricRecord(ctx, *s.taskOpts)
+	if err != nil {
+		return errors.Wrap(err, "problem creating cedar system metrics metadata object")
+	}
+
+	var streamingCtx context.Context
+	streamingCtx, s.streamingCancel = context.WithCancel(ctx)
+
+	for _, collector := range s.collectors {
+		stream, err := s.client.StreamSystemMetrics(streamingCtx, s.id, collector.Name(), metrics.StreamOpts{})
+		if err != nil {
+			s.streamingCancel()
+			s.streamingCancel = nil
+			return errors.Wrap(err, fmt.Sprintf("problem creating system metrics stream for id %s and metricType %s", s.id, collector.Name()))
+		}
+
+		ftdcCollector := ftdc.NewStreamingCollector(1, stream)
+		err = ftdcCollector.SetMetadata(collector.Metadata())
+		if err != nil {
+			s.streamingCancel()
+			s.streamingCancel = nil
+			return errors.Wrap(err, fmt.Sprintf("problem setting metadata on stream for id %s and metricType %s", s.id, collector.Name()))
+		}
+
+		s.wg.Add(1)
+		go s.timedCollect(streamingCtx, collector, stream, ftdcCollector)
+	}
+	go s.closeOnCancel(ctx, streamingCtx)
+
+	return nil
+}
+
+func (s *SystemMetricsCollector) timedCollect(ctx context.Context, mc MetricCollector, stream *metrics.SystemMetricsWriteCloser, collector ftdc.Collector) {
+	defer s.wg.Done()
+
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	defer s.catcher.Add(errors.Wrap(stream.Close(), fmt.Sprintf("problem closing system metrics stream for id %s and metricType %s", s.id, mc.Name())))
```

Comment:
Probably combine all the defers since having so many together makes it difficult to understand since they execute in LIFO order.